docs/custom-component-one-off.qmd

```diff
@@ -0,0 +1,323 @@
+    renderValue(el, payload) { ... }
+}
+
+// Register the binding
+Shiny.outputBindings.register(
+  new TabulatorOutputBinding(),
+  "shiny-tabulator-output"
+);
+```
+
+This class has two methods that we need to implement: `find()` and `renderValue()`. The `find()` method is used to find the element that we want to render the table in. The `renderValue()` method is used to render the table in the element. After making that class we need to register it with Shiny so it can find and send data to instances of our output.
+
+### The `find()` method
+
+Now that we have the scaffolding setup, we can fill it in. Starting with the `find` method, this function is passed a `scope` object, which is a `JQuery` selection and should return the element you wish to render your output into.
+
+```{.javascript filename="tableComponent.js"}
+class TabulatorOutputBinding extends Shiny.OutputBinding {
+    find(scope) {
+        return scope.find(".shiny-tabulator-output");
+    }
+
+    renderValue(el, payload) {...}
+}
+
+Shiny.outputBindings.register(...);
+```
+
+Note that we're using the class `".shiny-tabulator-output"` here to mark the element that we want to render the table in. This is the same class that we will use in our `output_tabulator()` function in our app's server code. You can use any valid css selector here, but it's common to use a descriptive class name like this.
+
+### The `renderValue()` method
+
+Next, we fill in the main logic for rendering our table in to the `renderValue` method. This method gets passed two arguments: `el` which is an HTMLElement as found by our find function, and `payload` which is the data that the server has provided from the render function (more on this [soon.](#the-render_tabulator-decorator))
+
+```{.javascript filename="tableComponent.js"}
+// Import the Tabulator library
+import { Tabulator } from "./tabulator_esm.min.js";
+
+class TabulatorOutputBinding extends Shiny.OutputBinding {
+    find(scope) { ... }
+
+    renderValue(el, payload) {
+      // Unpack the info we get from the associated render function
+      const { columns, data, type_hints } = payload;
+
+      // Convert the column names to a format that Tabulator expects
+      const columnsDef = columns.map((col, i) => {
+        return {
+          title: col,
+          field: col,
+          hozAlign: type_hints[i] === "numeric" ? "right" : "left",
+        };
+      });
+
+      // Data comes in as a series of rows with each row having as many elements
+      // as there are columns in the data. We need to map this to a series of
+      // objects with keys corresponding to the column names.
+      function zipRowWithColumns(row) {
+        const obj = {};
+        row.forEach((val, i) => {
+          obj[columns[i]] = val;
+        });
+        return obj;
+      }
+
+      // Render the table
+      new Tabulator(el, {
+        data: data.map(zipRowWithColumns),
+        layout: "fitColumns",
+        columns: columnsDef,
+      });
```

Comment:
I might be missing something, but how does simply instantiating the `Tabulator` object render it into the page?

If I am missing something, it might help to explain how that happens.

Reply:
Added a comment. You _can_ hold onto the object and make updates but we dont do that here. Just instantiating like this will render the table, though.